src/lxc/cgroups/cgfsng.c

```diff
@@ -1188,18 +1207,71 @@ static bool cgroup_tree_create(struct hierarchy *h, const char *cgroup_tree,
 
 static void cgroup_tree_leaf_remove(struct hierarchy *h, bool payload)
 {
-	__do_free char *full_path = NULL;
+	__do_free char *full_path = NULL, *__limit_path = NULL;
+	char *limit_path = NULL;
 
 	if (payload) {
 		__lxc_unused __do_close int fd = move_fd(h->cgfd_con);
 		full_path = move_ptr(h->container_full_path);
+		limit_path = move_ptr(h->container_limit_path);
+		if (limit_path != full_path)
+			__limit_path = limit_path;
 	} else {
 		__lxc_unused __do_close int fd = move_fd(h->cgfd_mon);
 		full_path = move_ptr(h->monitor_full_path);
 	}
 
 	if (full_path && rmdir(full_path))
 		SYSWARN("Failed to rmdir(\"%s\") cgroup", full_path);
+	if (limit_path && rmdir(limit_path))
+		SYSWARN("Failed to rmdir(\"%s\") cgroup", limit_path);
+}
+
+/*
+ * Check we have no lxc.cgroup.dir, and that lxc.cgroup.dir.limit_prefix is a
+ * proper prefix directory of lxc.cgroup.dir.payload.
+ *
+ * Returns the prefix length if it is set, otherwise zero on success.
+ */
+static bool check_cgroup_dir_config(struct lxc_conf *conf)
+{
+	const char *payload_dir, *limit_prefix, *monitor_dir;
+	size_t prefix_len;
+
+	monitor_dir = conf->cgroup_meta.monitor_dir;
+	payload_dir = conf->cgroup_meta.payload_dir;
+	limit_prefix = conf->cgroup_meta.limit_prefix_dir;
+
+	/* none of the new options are set, all is fine */
+	if (!monitor_dir && !payload_dir && !limit_prefix)
+		return true;
+
+	/* some are set, make sure lxc.cgroup.dir is not also set*/
+	if (conf->cgroup_meta.dir)
+		return log_error_errno(false, EINVAL,
+				       "lxc.cgroup.dir conflicts with"
+				       " lxc.cgroup.dir.payload/monitor");
+
+	/* make sure both monitor and payload are set */
+	if (!monitor_dir || !payload_dir)
+		return log_error_errno(false, EINVAL,
+				       "lxc.cgroup.dir.payload and"
+				       " lxc.cgroup.dir.monitor must both be"
+				       " set");
+
+	if (!limit_prefix)
+		return true;
+
+	/* we have a limit prefix directory, check it: */
+
+	prefix_len = strlen(limit_prefix);
+	if (strncmp(payload_dir, limit_prefix, prefix_len) == 0 &&
+	    payload_dir[prefix_len] == '/')
```

Comment:
Any concerns about having `../../../` or sm in the path?

Reply:
Maybe we just just make this a hard requirement for all `lxc.cgroup.{dir,monitor,payload,limit_prefix}`?

Reply:
Not sure, if so, `lxc.cgroup.dir` should be restricted as well. I'm not sure it's our job to check this here though. An unprivileged user shouldn't have the permissions for "bad paths" anyway.

Reply:
Alternatively we could go with `payload_dir` and `namespace_dir` (closer to the old patch), that might also be easier in some places (eg. I already need to take the special care of the cpuset cgroup which I just realized wouldn't work with deeper paths than 1 additional dir).